idea-plugin/src/main/kotlin/arrow/meta/ide/phases/resolve/QuoteSystemCache.kt

```diff
@@ -0,0 +1,388 @@
+import com.intellij.openapi.roots.FileIndexFacade
+import com.intellij.openapi.startup.StartupManager
+import com.intellij.openapi.vfs.AsyncFileListener
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.openapi.vfs.VirtualFileManager
+import com.intellij.openapi.vfs.newvfs.events.VFileContentChangeEvent
+import com.intellij.openapi.vfs.newvfs.events.VFileCopyEvent
+import com.intellij.openapi.vfs.newvfs.events.VFileEvent
+import com.intellij.openapi.vfs.newvfs.events.VFileMoveEvent
+import com.intellij.psi.PsiDocumentManager
+import com.intellij.psi.PsiManager
+import com.intellij.psi.search.FileTypeIndex
+import org.jetbrains.annotations.TestOnly
+import org.jetbrains.kotlin.caches.resolve.KotlinCacheService
+import org.jetbrains.kotlin.cli.common.messages.CompilerMessageLocation
+import org.jetbrains.kotlin.cli.common.messages.CompilerMessageSeverity
+import org.jetbrains.kotlin.cli.common.messages.MessageCollector
+import org.jetbrains.kotlin.descriptors.DeclarationDescriptor
+import org.jetbrains.kotlin.idea.KotlinFileType
+import org.jetbrains.kotlin.idea.debugger.readAction
+import org.jetbrains.kotlin.idea.search.projectScope
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+import org.jetbrains.kotlin.resolve.lazy.LazyEntity
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicLong
+
+/**
+ * QuoteSystemCache is a project component which manages the transformations of KtFiles by the quote system.
+ *
+ * When initialized, it transforms all .kt files of the project in a background thread.
+ */
+class QuoteSystemCache(private val project: Project) : ProjectComponent, Disposable {
+  companion object {
+    fun getInstance(project: Project): QuoteSystemCache = project.getComponent(QuoteSystemCache::class.java)
+
+    /**
+     * messages is used to report messages generated by quote system via IntelliJ's log.
+     */
+    private val messages: MessageCollector = object : MessageCollector {
+      override fun clear() {}
+
+      override fun hasErrors(): Boolean = false
+
+      override fun report(severity: CompilerMessageSeverity, message: String, location: CompilerMessageLocation?) {
+        when {
+          severity.isError -> LOG.error(message)
+          severity.isWarning -> LOG.warn(message)
+          else -> LOG.debug(message)
+        }
+      }
+    }
+  }
+
+  private val metaCache: MetaTransformationCache = DefaultMetaTransformationCache()
+
+  // pool where the quote system transformations are executed.
+  // this is a single thread pool to avoid concurrent updates to the cache.
+  // we keep a pool per project, so that we're able to shut it down when the project is closed
+  private val pool: ExecutorService = Executors.newFixedThreadPool(1) { runnable -> Thread(runnable, "arrow worker, ${project.name}") }
```

Comment:
Do submitted running and on hold tasks get automatically canceled if a user action produces invalidation or requires a fresh computation of the cache ?